internal/certdb/certdb.go

```diff
@@ -0,0 +1,115 @@
+package certdb
+
+import (
+	"database/sql"
+	"fmt"
+
+	_ "github.com/mattn/go-sqlite3"
+)
+
+const queryCreateTable = "CREATE TABLE IF NOT EXISTS %s (CSR VARCHAR PRIMARY KEY UNIQUE NOT NULL, Certificate VARCHAR)"
+
+const queryGetAllCSRs = "SELECT rowid, * FROM %s"
+const queryGetCSR = "SELECT rowid, * FROM %s WHERE CSR=?"
+const queryCreateCSR = "INSERT INTO %s (CSR) VALUES (?)"
+const queryUpdateCSR = "UPDATE %s SET Certificate=? WHERE CSR=?"
+const queryDeleteCSR = "DELETE FROM %s WHERE CSR=?"
+
+type CertificateRequests struct {
+	table string
+	conn  *sql.DB
+}
+
+type CertificateRequest struct {
+	ID          int
+	CSR         string
+	Certificate *string
+}
+
+func (db *CertificateRequests) RetrieveAll() ([]*CertificateRequest, error) {
+	rows, err := db.conn.Query(fmt.Sprintf(queryGetAllCSRs, db.table))
+	if err != nil {
+		return nil, err
+	}
+
+	var allCsrs []*CertificateRequest
+	defer rows.Close()
+	for rows.Next() {
+		var csr CertificateRequest
+		if err := rows.Scan(&csr.ID, &csr.CSR, &csr.Certificate); err != nil {
+			return nil, err
+		}
+		allCsrs = append(allCsrs, &csr)
+	}
+	return allCsrs, nil
+}
+
+func (db *CertificateRequests) Retrieve(csr *string) (*CertificateRequest, error) {
```

Comment:
Is there a reason I am missing why the parameter is a pointer? Usually basic types are passed by value in Go, as they are cheap to copy.

Reply:
This comment is valid for all the function parameters of `*string`.

Reply:
Pointers are all gone.